test/ruby/test_syntax.rb

```diff
@@ -832,6 +832,46 @@ def test_parenthesised_statement_argument
     assert_valid_syntax("foo (bar rescue nil)")
   end
 
+  def test_regex_heredoc_x
+    expected = eval("/a\n/x")
+    result = eval("<</eos/x\na\neos")
+    assert_equal(expected, result)
+  end
+
+  def test_regex_heredoc
+    expected = eval("/a\n/")
+    result = eval("<</eos/\na\neos")
+    assert_equal(expected, result)
+  end
+
+  def test_regex_indent_heredoc
+    expected = eval("/  a\n/x")
+    result = eval("  <<-/eos/x\n  a\n  eos")
+    assert_equal(expected, result)
+  end
+
+  def test_regex_deindent_heredoc
+    expected = eval("/a\n b\nc\n/x")
+    result = eval("<<~/eos/x\n  a\n   b\n  c\neos")
+    assert_equal(expected, result)
+  end
+
+  def test_regex_deindent_indented_heredoc
+    expected = eval("/a\n b\nc\n/x")
+    result = eval("  <<~/eos/x\n    a\n     b\n    c\n  eos")
+    assert_equal(expected, result)
+  end
+
+  def test_regex_deindent_indented2_heredoc
+    expected = eval("/  a\n   b\nc\n/x")
+    result = eval("  <<~/eos/x\n      a\n       b\n    c\n  eos")
+    assert_equal(expected, result)
+  end
+
+  def test_regex_heredoc_bad_switch
+    assert_raise(SyntaxError, "bad switch") { eval("<</eos/xz\nbob\neos") }
+  end
+
```

Comment:
There is very little difference in how the string/xstring cases are handled and the regexp case, so I don't think a lot of testing is needed for `__LINE__` etc.  But I am open to suggestions.